lambdas/indexer/app.py

```diff
@@ -231,95 +225,52 @@ def index(event: chalice.app.SQSEvent):
-            indexer.aggregate(tallies)
-
-        if deferrals:
-            for tally in deferrals:
-                log.info('Deferring aggregation of %i contribution(s) to entity %s/%s',
-                         tally.num_contributions, tally.entity.entity_type, tally.entity.entity_id)
-            document_queue.send_messages(Entries=[dict(tally.to_message(), Id=str(i))
-                                                  for i, tally in enumerate(deferrals)])
-
-        document_queue.delete_messages(Entries=[dict(Id=str(i), ReceiptHandle=message.receipt_handle)
-                                                for i, message in enumerate(chain(messages))])
-        total -= len(messages)
-        total += len(deferrals)
-    else:
-        tokens, expired_tokens = map(list, partition(Token.expired, tokens))
-        if expired_tokens:
-            expired_total = sum(token.value for token in expired_tokens)
-            log.info('Expiring %i token(s) with a total value of %i', len(expired_tokens), expired_total)
-            total -= expired_total
-
-    if total > 0:
-        expiration = max(token.expiration for token in tokens)
-        tokens = Token.mint_many(total, expiration=expiration)
-        delay = 0 if messages else round(random.uniform(30, 90))
-        log.info('Recirculating %i token(s), after a delay of %is, for a total value of %i, expiring at %s',
-                 len(tokens), delay, total, datetime.utcfromtimestamp(expiration).isoformat(timespec='seconds') + 'Z')
-        _send_tokens(tokens, delay=delay)
-    elif total < 0:
-        # Discarding token deficit will lead to an overall surplus of token value which will expire at some point.
-        log.info("Discarding token deficit of %i.", total)
+
+    # Consolidate multiple tallies for the same entity and process entities with only one message. Because SQS FIFO
+    # queues try to put as many messages from the same message group in a reception batch, a single message per
+    # group may indicate that that message is the last one in the group. Inversely, multiple messages per group
+    # in a batch are a likely indicator for the presence of even more queued messages in that group. The more
+    # bundle contributions we defer, the higher the amortized savings on aggregation become. Aggregating bundle
+    # contributions is a costly operation for any entity with many contributions e.g., a large project.
+    #
+    tallies_by_id: MutableMapping[EntityReference, List[DocumentTally]] = defaultdict(list)
+    for record in event['Records']:
+        tally = DocumentTally.from_message(record)
+        log.info('Attempt %i of handling %i contribution(s) for entity %s/%s',
+                 tally.attempts, tally.num_contributions, tally.entity.entity_type, tally.entity.entity_id)
+        tallies_by_id[tally.entity].append(tally)
+        # NOTE here we are grouping the tallies based on their entity
```

Comment:
I think this comment is unnecessary since this information can be quickly deduced by looking at the type annotation for `tallies_by_id`.

Reply:
whoops, that was a personal note.